## Import 

### Import Libraries

In [128]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

# setting seaborn style
sns.set()

from sklearn.model_selection import train_test_split

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_colwidth', None) #Show all columns


### Import Datasets

In [129]:
train_data = pd.read_csv("project_data/train_data.csv", low_memory=False)
test_data = pd.read_csv("project_data/test_data.csv")

In [130]:
print(f'Number of rows and columns in train_data: {train_data.shape}')
print(f'Number of rows and columns in test_data: {train_data.shape}')

Number of rows and columns in train_data: (593471, 33)
Number of rows and columns in test_data: (593471, 33)


In [131]:
train_data.head()

Accident Date  Age at Injury Alternative Dispute Resolution Assembly Date  \
0    2019-12-30           31.0                              N    2020-01-01   
1    2019-08-30           46.0                              N    2020-01-01   
2    2019-12-06           40.0                              N    2020-01-01   
3           NaN            NaN                            NaN    2020-01-01   
4    2019-12-30           61.0                              N    2020-01-01   

  Attorney/Representative  Average Weekly Wage  Birth Year    C-2 Date  \
0                       N                 0.00      1988.0  2019-12-31   
1                       Y              1745.93      1973.0  2020-01-01   
2                       N              1434.80      1979.0  2020-01-01   
3                     NaN                  NaN         NaN         NaN   
4                       N                  NaN      1958.0  2019-12-31   

     C-3 Date                  Carrier Name  ... WCIO Cause of Injury Code  \
0         NaN    NEW HAMPSHIRE INSURANCE CO  ...                      27.0   
1  2020-01-14  ZURICH AMERICAN INSURANCE CO  ...                      97.0   
2         NaN     INDEMNITY INSURANCE CO OF  ...                      79.0   
3         NaN                           NaN  ...                       NaN   
4         NaN          STATE INSURANCE FUND  ...                      16.0   

   WCIO Cause of Injury Description WCIO Nature of Injury Code  \
0      FROM LIQUID OR GREASE SPILLS                       10.0   
1                 REPETITIVE MOTION                       49.0   
2    OBJECT BEING LIFTED OR HANDLED                        7.0   
3                               NaN                        NaN   
4   HAND TOOL, UTENSIL; NOT POWERED                       43.0   

  WCIO Nature of Injury Description WCIO Part Of Body Code  \
0                         CONTUSION                   62.0   
1                    SPRAIN OR TEAR                   38.0   
2                        CONCUSSION                   10.0   
3                               NaN                    NaN   
4                          PUNCTURE                   36.0   

  WCIO Part Of Body Description Zip Code Agreement Reached      WCB Decision  \
0                      BUTTOCKS    13662               0.0  Not Work Related   
1                   SHOULDER(S)    14569               1.0  Not Work Related   
2          MULTIPLE HEAD INJURY    12589               0.0  Not Work Related   
3                           NaN      NaN               NaN               NaN   
4                     FINGER(S)    12603               0.0  Not Work Related   

   Number of Dependents  
0                   1.0  
1                   4.0  
2                   6.0  
3                   NaN  
4                   1.0  

[5 rows x 33 columns]

In [132]:
test_data.head()

Accident Date  Age at Injury Alternative Dispute Resolution Assembly Date  \
0    2022-12-24             19                              N    2023-01-02   
1    2022-11-20             19                              N    2023-01-02   
2    2022-12-26             59                              N    2023-01-02   
3    2022-12-28             55                              N    2023-01-02   
4    2022-12-20             25                              N    2023-01-02   

  Attorney/Representative  Average Weekly Wage  Birth Year    C-2 Date  \
0                       N                  NaN      2003.0  2023-01-02   
1                       N                  NaN      2003.0  2023-01-02   
2                       N                  0.0      1963.0  2022-12-31   
3                       N                  0.0         0.0  2023-01-02   
4                       N                  0.0      1997.0  2022-12-31   

  C-3 Date                 Carrier Name  ... Medical Fee Region  \
0      NaN    INDEMNITY INSURANCE CO OF  ...                 IV   
1      NaN      A I U INSURANCE COMPANY  ...                 IV   
2      NaN    AMGUARD INSURANCE COMPANY  ...                III   
3      NaN  INDEMNITY INS. OF N AMERICA  ...                 IV   
4      NaN   NEW HAMPSHIRE INSURANCE CO  ...                 IV   

   OIICS Nature of Injury Description WCIO Cause of Injury Code  \
0                                 NaN                      31.0   
1                                 NaN                      75.0   
2                                 NaN                      68.0   
3                                 NaN                      25.0   
4                                 NaN                      79.0   

   WCIO Cause of Injury Description WCIO Nature of Injury Code  \
0           FALL, SLIP OR TRIP, NOC                       10.0   
1          FALLING OR FLYING OBJECT                       10.0   
2                 STATIONARY OBJECT                       49.0   
3  FROM DIFFERENT LEVEL (ELEVATION)                       10.0   
4    OBJECT BEING LIFTED OR HANDLED                       40.0   

  WCIO Nature of Injury Description WCIO Part Of Body Code  \
0                         CONTUSION                   54.0   
1                         CONTUSION                   10.0   
2                    SPRAIN OR TEAR                   62.0   
3                         CONTUSION                   53.0   
4                        LACERATION                   37.0   

   WCIO Part Of Body Description  Zip Code Number of Dependents  
0                      LOWER LEG     10466                    1  
1           MULTIPLE HEAD INJURY     11691                    1  
2                       BUTTOCKS     10604                    0  
3                           KNEE     11411                    6  
4                          THUMB     11212                    5  

[5 rows x 30 columns]

In [133]:
train_data.tail()

Accident Date  Age at Injury Alternative Dispute Resolution  \
593466           NaN            NaN                            NaN   
593467    2022-12-13           72.0                              N   
593468           NaN            NaN                            NaN   
593469           NaN            NaN                            NaN   
593470           NaN            NaN                            NaN   

       Assembly Date Attorney/Representative  Average Weekly Wage  Birth Year  \
593466    2022-12-31                     NaN                  NaN         NaN   
593467    2022-12-31                       N                  0.0      1950.0   
593468    2022-12-31                     NaN                  NaN         NaN   
593469    2022-12-31                     NaN                  NaN         NaN   
593470    2022-12-31                     NaN                  NaN         NaN   

          C-2 Date C-3 Date                   Carrier Name  ...  \
593466         NaN      NaN                            NaN  ...   
593467  2022-12-31      NaN  TECHNOLOGY INSURANCE CO. INC.  ...   
593468         NaN      NaN                            NaN  ...   
593469         NaN      NaN                            NaN  ...   
593470         NaN      NaN                            NaN  ...   

       WCIO Cause of Injury Code  WCIO Cause of Injury Description  \
593466                       NaN                               NaN   
593467                      25.0  FROM DIFFERENT LEVEL (ELEVATION)   
593468                       NaN                               NaN   
593469                       NaN                               NaN   
593470                       NaN                               NaN   

       WCIO Nature of Injury Code WCIO Nature of Injury Description  \
593466                        NaN                               NaN   
593467                       90.0   MULTIPLE PHYSICAL INJURIES ONLY   
593468                        NaN                               NaN   
593469                        NaN                               NaN   
593470                        NaN                               NaN   

       WCIO Part Of Body Code WCIO Part Of Body Description Zip Code  \
593466                    NaN                           NaN      NaN   
593467                   -9.0                      MULTIPLE    12779   
593468                    NaN                           NaN      NaN   
593469                    NaN                           NaN      NaN   
593470                    NaN                           NaN      NaN   

       Agreement Reached      WCB Decision  Number of Dependents  
593466               NaN               NaN                   NaN  
593467               0.0  Not Work Related                   3.0  
593468               NaN               NaN                   NaN  
593469               NaN               NaN                   NaN  
593470               NaN               NaN                   NaN  

[5 rows x 33 columns]

In [134]:
test_data.tail()

Accident Date  Age at Injury Alternative Dispute Resolution  \
387970    2012-09-12             52                              N   
387971    2024-05-22             59                              N   
387972    2024-05-06             45                              N   
387973    2024-02-24             42                              N   
387974           NaN              0                              N   

       Assembly Date Attorney/Representative  Average Weekly Wage  Birth Year  \
387970    2024-06-05                       N                  NaN      1960.0   
387971    2024-06-05                       Y                  0.0      1965.0   
387972    2024-06-05                       Y                  0.0      1979.0   
387973    2024-06-05                       Y                  0.0      1981.0   
387974    2024-06-05                       Y                  0.0         NaN   

          C-2 Date    C-3 Date                   Carrier Name  ...  \
387970  2012-10-23         NaN           STATE INSURANCE FUND  ...   
387971         NaN  2024-05-28          NYC TRANSIT AUTHORITY  ...   
387972         NaN         NaN           STATE INSURANCE FUND  ...   
387973         NaN  2024-05-21  NEW YORK BLACK CAR OPERATORS'  ...   
387974         NaN  2024-05-28   CNY OTHER THAN ED, HED WATER  ...   

       Medical Fee Region  OIICS Nature of Injury Description  \
387970                  I                                 NaN   
387971                 IV                                 NaN   
387972                 IV                                 NaN   
387973                 IV                                 NaN   
387974                 IV                                 NaN   

       WCIO Cause of Injury Code WCIO Cause of Injury Description  \
387970                       NaN                              NaN   
387971                       NaN                              NaN   
387972                       NaN                              NaN   
387973                       NaN                              NaN   
387974                       NaN                              NaN   

       WCIO Nature of Injury Code WCIO Nature of Injury Description  \
387970                        NaN                               NaN   
387971                        NaN                               NaN   
387972                        NaN                               NaN   
387973                        NaN                               NaN   
387974                        NaN                               NaN   

       WCIO Part Of Body Code  WCIO Part Of Body Description  Zip Code  \
387970                    NaN                            NaN     13619   
387971                    NaN                            NaN     11776   
387972                    NaN                            NaN     11368   
387973                    NaN                            NaN     11432   
387974                    NaN                            NaN     10302   

       Number of Dependents  
387970                    5  
387971                    1  
387972                    5  
387973                    5  
387974                    5  

[5 rows x 30 columns]

### Descriptive analysis

In [135]:
# set "Claim Identifier" as the index for the train and test datasets
train_data = train_data.set_index("Claim Identifier")
test_data = test_data.set_index("Claim Identifier")

In [136]:
train_data.head()

Accident Date  Age at Injury Alternative Dispute Resolution  \
Claim Identifier                                                               
5393875             2019-12-30           31.0                              N   
5393091             2019-08-30           46.0                              N   
5393889             2019-12-06           40.0                              N   
957648180                  NaN            NaN                            NaN   
5393887             2019-12-30           61.0                              N   

                 Assembly Date Attorney/Representative  Average Weekly Wage  \
Claim Identifier                                                              
5393875             2020-01-01                       N                 0.00   
5393091             2020-01-01                       Y              1745.93   
5393889             2020-01-01                       N              1434.80   
957648180           2020-01-01                     NaN                  NaN   
5393887             2020-01-01                       N                  NaN   

                  Birth Year    C-2 Date    C-3 Date  \
Claim Identifier                                       
5393875               1988.0  2019-12-31         NaN   
5393091               1973.0  2020-01-01  2020-01-14   
5393889               1979.0  2020-01-01         NaN   
957648180                NaN         NaN         NaN   
5393887               1958.0  2019-12-31         NaN   

                                  Carrier Name  ... WCIO Cause of Injury Code  \
Claim Identifier                                ...                             
5393875             NEW HAMPSHIRE INSURANCE CO  ...                      27.0   
5393091           ZURICH AMERICAN INSURANCE CO  ...                      97.0   
5393889              INDEMNITY INSURANCE CO OF  ...                      79.0   
957648180                                  NaN  ...                       NaN   
5393887                   STATE INSURANCE FUND  ...                      16.0   

                 WCIO Cause of Injury Description WCIO Nature of Injury Code  \
Claim Identifier                                                               
5393875              FROM LIQUID OR GREASE SPILLS                       10.0   
5393091                         REPETITIVE MOTION                       49.0   
5393889            OBJECT BEING LIFTED OR HANDLED                        7.0   
957648180                                     NaN                        NaN   
5393887           HAND TOOL, UTENSIL; NOT POWERED                       43.0   

                 WCIO Nature of Injury Description WCIO Part Of Body Code  \
Claim Identifier                                                            
5393875                                  CONTUSION                   62.0   
5393091                             SPRAIN OR TEAR                   38.0   
5393889                                 CONCUSSION                   10.0   
957648180                                      NaN                    NaN   
5393887                                   PUNCTURE                   36.0   

                 WCIO Part Of Body Description Zip Code  Agreement Reached  \
Claim Identifier                                                             
5393875                               BUTTOCKS    13662                0.0   
5393091                            SHOULDER(S)    14569                1.0   
5393889                   MULTIPLE HEAD INJURY    12589                0.0   
957648180                                  NaN      NaN                NaN   
5393887                              FINGER(S)    12603                0.0   

                      WCB Decision Number of Dependents  
Claim Identifier                                         
5393875           Not Work Related                  1.0  
5393091           Not Work Related                  4.0  
5393889           Not Work Related                  6.0  
957

In [137]:
train_data.columns.values

array(['Accident Date', 'Age at Injury', 'Alternative Dispute Resolution',
       'Assembly Date', 'Attorney/Representative', 'Average Weekly Wage',
       'Birth Year', 'C-2 Date', 'C-3 Date', 'Carrier Name',
       'Carrier Type', 'Claim Injury Type', 'County of Injury',
       'COVID-19 Indicator', 'District Name', 'First Hearing Date',
       'Gender', 'IME-4 Count', 'Industry Code',
       'Industry Code Description', 'Medical Fee Region',
       'OIICS Nature of Injury Description', 'WCIO Cause of Injury Code',
       'WCIO Cause of Injury Description', 'WCIO Nature of Injury Code',
       'WCIO Nature of Injury Description', 'WCIO Part Of Body Code',
       'WCIO Part Of Body Description', 'Zip Code', 'Agreement Reached',
       'WCB Decision', 'Number of Dependents'], dtype=object)

In [138]:
train_data.describe(include="all").T

count unique  \
Accident Date                         570337   5539   
Age at Injury                       574026.0    NaN   
Alternative Dispute Resolution        574026      3   
Assembly Date                         593471   1096   
Attorney/Representative               574026      2   
Average Weekly Wage                 545375.0    NaN   
Birth Year                          544948.0    NaN   
C-2 Date                              559466   2475   
C-3 Date                              187245   1648   
Carrier Name                          574026   2046   
Carrier Type                          574026      8   
Claim Injury Type                     574026      8   
County of Injury                      574026     63   
COVID-19 Indicator                    574026      2   
District Name                         574026      8   
First Hearing Date                    150798   1094   
Gender                                574026      4   
IME-4 Count                         132803.0    NaN   
Industry Code                       564068.0    NaN   
Industry Code Description             564068     20   
Medical Fee Region                    574026      5   
OIICS Nature of Injury Description       0.0    NaN   
WCIO Cause of Injury Code           558386.0    NaN   
WCIO Cause of Injury Description      558386     74   
WCIO Nature of Injury Code          558369.0    NaN   
WCIO Nature of Injury Description     558369     56   
WCIO Part Of Body Code              556944.0    NaN   
WCIO Part Of Body Description         556944     54   
Zip Code                              545389   8286   
Agreement Reached                   574026.0    NaN   
WCB Decision                          574026      1   
Number of Dependents                574026.0    NaN   

                                                                  top    freq  \
Accident Date                                              2020-03-01    1245   
Age at Injury                                                     NaN     NaN   
Alternative Dispute Resolution                                      N  571412   
Assembly Date                                              2020-03-06    1422   
Attorney/Representative                                             N  392291   
Average Weekly Wage                                               NaN     NaN   
Birth Year                                                        NaN     NaN   
C-2 Date                                                   2021-05-11    1847   
C-3 Date                                                   2021-04-21     350   
Carrier Name                                     STATE INSURANCE FUND  111144   
Carrier Type                                              1A. PRIVATE  285368   
Claim Injury Type                                         2. NON-COMP  291078   
County of Injury                                              SUFFOLK   60430   
COVID-19 Indicator                                                  N  546505   
District Name                                                     NYC  270779   
First Hearing Date                                         2022-04-15     440   
Gender                                                              M  335218   
IME-4 Count                                                       NaN     NaN   
Industry Code                                                     NaN     NaN   
Industry Code Description           HEALTH CARE AND SOCIAL ASSISTANCE  114339   
Medical Fee Region                                                 IV  265981   
OIICS Nature of Injury Description                                NaN     NaN   
WCIO Cause of Injury Code                                         NaN     NaN   
WCIO Cause of Injury Description                              LIFTING   46610   
WCIO Nature of Injury Code                                        NaN     NaN   
WCIO Nature of Injury Description                      STRAIN OR TEAR  153373   
WCIO Part Of Body Code              

In [139]:
train_data.info()

## DATA TYPES:
# Convert accident date (to date type)
# convert age (to int)
# Alternative dispute resolution (??)
# Assembly date (to date type)
# birth year (to int)
# C2 date (to date type)
# C3 date (to date type)
# First hearing date (to date)
# IME-4 count (to int)
# Industry code (object)
# OIICS Nature of Injury Description (to object) - all nulls (relevant??)
# WCIO cause of injury code (float to object)
# WCIO Nature of Injury Code (float to object)
# WCIO Part Of Body Code (float to object)
# Agreement reached (??) - boolean/ int(???)
# Number of dependents (to int)

<class 'pandas.core.frame.DataFrame'>
Index: 593471 entries, 5393875 to 818961390
Data columns (total 32 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   Accident Date                       570337 non-null  object 
 1   Age at Injury                       574026 non-null  float64
 2   Alternative Dispute Resolution      574026 non-null  object 
 3   Assembly Date                       593471 non-null  object 
 4   Attorney/Representative             574026 non-null  object 
 5   Average Weekly Wage                 545375 non-null  float64
 6   Birth Year                          544948 non-null  float64
 7   C-2 Date                            559466 non-null  object 
 8   C-3 Date                            187245 non-null  object 
 9   Carrier Name                        574026 non-null  object 
 10  Carrier Type                        574026 non-null  object 
 11  Claim Injury Type     

#### Check for duplicates & missing values in target variable:

In [140]:
'''
Julia comment: I think check for duplicates should be in the beggining: after setting the index, 
also maybe it's better to use train_data.duplicated().sum() (it is easier to read)
And then after checking the number of duplicates i would just drop all of them
so train_data.drop_duplicates(inplace=True)

'''
# duplicates =train_data.duplicated().value_counts()
# duplicates

#Julia comment
train_data.duplicated().sum()
train_data.drop_duplicates(inplace=True)

In [141]:
train_data["Claim Injury Type"].unique()

array(['2. NON-COMP', '4. TEMPORARY', nan, '3. MED ONLY',
       '5. PPD SCH LOSS', '6. PPD NSL', '1. CANCELLED', '8. DEATH',
       '7. PTD'], dtype=object)

In [142]:
missing_data_target = train_data[train_data['Claim Injury Type'].isna()]
missing_data_target.head()

Accident Date  Age at Injury Alternative Dispute Resolution  \
Claim Identifier                                                               
957648180                  NaN            NaN                            NaN   
653337074                  NaN            NaN                            NaN   
489154817                  NaN            NaN                            NaN   
351928125                  NaN            NaN                            NaN   
454184427                  NaN            NaN                            NaN   

                 Assembly Date Attorney/Representative  Average Weekly Wage  \
Claim Identifier                                                              
957648180           2020-01-01                     NaN                  NaN   
653337074           2020-01-02                     NaN                  NaN   
489154817           2020-01-03                     NaN                  NaN   
351928125           2020-01-04                     NaN                  NaN   
454184427           2020-01-05                     NaN                  NaN   

                  Birth Year C-2 Date C-3 Date Carrier Name  ...  \
Claim Identifier                                             ...   
957648180                NaN      NaN      NaN          NaN  ...   
653337074                NaN      NaN      NaN          NaN  ...   
489154817                NaN      NaN      NaN          NaN  ...   
351928125                NaN      NaN      NaN          NaN  ...   
454184427                NaN      NaN      NaN          NaN  ...   

                 WCIO Cause of Injury Code WCIO Cause of Injury Description  \
Claim Identifier                                                              
957648180                              NaN                              NaN   
653337074                              NaN                              NaN   
489154817                              NaN                              NaN   
351928125                              NaN                              NaN   
454184427                              NaN                              NaN   

                 WCIO Nature of Injury Code WCIO Nature of Injury Description  \
Claim Identifier                                                                
957648180                               NaN                               NaN   
653337074                               NaN                               NaN   
489154817                               NaN                               NaN   
351928125                               NaN                               NaN   
454184427                               NaN                               NaN   

                 WCIO Part Of Body Code WCIO Part Of Body Description  \
Claim Identifier                                                        
957648180                           NaN                           NaN   
653337074                           NaN                           NaN   
489154817                           NaN                           NaN   
351928125                           NaN                           NaN   
454184427                           NaN                           NaN   

                 Zip Code  Agreement Reached  WCB Decision  \
Claim Identifier                                             
957648180             NaN                NaN           NaN   
653337074             NaN                NaN           NaN   
489154817             NaN                NaN           NaN   
351928125             NaN                NaN           NaN   
454184427             NaN                NaN           NaN   

                 Number of Dependents  
Claim Identifier                       
957648180                         NaN  
653337074                         NaN  
489154817                         NaN  
351928125                         NaN  
454184427                         NaN  

[5 rows x 32 columns]

In [143]:
missing_data_target.info()
# all features with missing values on the target, are also missing, except Assembly Date -> not relevant (decision -> drop rows)

<class 'pandas.core.frame.DataFrame'>
Index: 1096 entries, 957648180 to 208342104
Data columns (total 32 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Accident Date                       0 non-null      object 
 1   Age at Injury                       0 non-null      float64
 2   Alternative Dispute Resolution      0 non-null      object 
 3   Assembly Date                       1096 non-null   object 
 4   Attorney/Representative             0 non-null      object 
 5   Average Weekly Wage                 0 non-null      float64
 6   Birth Year                          0 non-null      float64
 7   C-2 Date                            0 non-null      object 
 8   C-3 Date                            0 non-null      object 
 9   Carrier Name                        0 non-null      object 
 10  Carrier Type                        0 non-null      object 
 11  Claim Injury Type                  

In [144]:
train_data_cleaned = train_data.drop(missing_data_target.index)

## Julia
### Missing values and strange answers in independent variables  

In [145]:
missing_values = train_data_cleaned.isnull().sum().sort_values(ascending=False)
print(missing_values)

OIICS Nature of Injury Description    574025
IME-4 Count                           441222
First Hearing Date                    423228
C-3 Date                              386781
Birth Year                             29078
Average Weekly Wage                    28651
Zip Code                               28637
WCIO Part Of Body Code                 17081
WCIO Part Of Body Description          17081
WCIO Nature of Injury Description      15656
WCIO Nature of Injury Code             15656
WCIO Cause of Injury Description       15639
WCIO Cause of Injury Code              15639
C-2 Date                               14559
Industry Code                           9957
Industry Code Description               9957
Accident Date                           3688
Medical Fee Region                         0
Agreement Reached                          0
WCB Decision                               0
Gender                                     0
Age at Injury                              0
District N

In [146]:
missing_percentage = ((train_data_cleaned.isnull().sum() / len(train_data_cleaned)) * 100).sort_values(ascending=False)
print(missing_percentage)

OIICS Nature of Injury Description    100.000000
IME-4 Count                            76.864596
First Hearing Date                     73.729890
C-3 Date                               67.380515
Birth Year                              5.065633
Average Weekly Wage                     4.991246
Zip Code                                4.988807
WCIO Part Of Body Code                  2.975654
WCIO Part Of Body Description           2.975654
WCIO Nature of Injury Description       2.727407
WCIO Nature of Injury Code              2.727407
WCIO Cause of Injury Description        2.724446
WCIO Cause of Injury Code               2.724446
C-2 Date                                2.536301
Industry Code                           1.734593
Industry Code Description               1.734593
Accident Date                           0.642481
Medical Fee Region                      0.000000
Agreement Reached                       0.000000
WCB Decision                            0.000000
Gender              

##### Drop 'OIICS Nature of Injury Description' feature & rows with missing target values 

In [151]:
# OIICS Nature of Injury Description containts only NaN values thats why we decided to drop that column 
train_data_cleaned =train_data.drop(['OIICS Nature of Injury Description'], axis=1)

In [147]:
columns_with_nan = missing_percentage[missing_percentage > 0].index.tolist()
columns_with_nan.remove('OIICS Nature of Injury Description')
columns_with_nan

['IME-4 Count',
 'First Hearing Date',
 'C-3 Date',
 'Birth Year',
 'Average Weekly Wage',
 'Zip Code',
 'WCIO Part Of Body Code',
 'WCIO Part Of Body Description',
 'WCIO Nature of Injury Description',
 'WCIO Nature of Injury Code',
 'WCIO Cause of Injury Description',
 'WCIO Cause of Injury Code',
 'C-2 Date',
 'Industry Code',
 'Industry Code Description',
 'Accident Date']

In [148]:
WCIO_columns = [
    'WCIO Part Of Body Code', 
    'WCIO Part Of Body Description', 
    'WCIO Nature of Injury Description', 
    'WCIO Nature of Injury Code', 
    'WCIO Cause of Injury Description', 
    'WCIO Cause of Injury Code'
]

# Display rows where all WCIO columns contain NaN values
WCIO_nan_rows = train_data_cleaned[train_data_cleaned[WCIO_columns].isna().all(axis=1)]

count_all_nan_WCIO = train_data_cleaned[WCIO_columns].isna().all(axis=1).sum()
print("Rows with all NaN values:", count_all_nan_WCIO/len(train_data_cleaned)*100)

WCIO_nan_rows.head()

Rows with all NaN values: 2.571839205609512


Accident Date  Age at Injury Alternative Dispute Resolution  \
Claim Identifier                                                               
5393991             2019-12-05           32.0                              N   
5394018             2019-12-11           44.0                              N   
5394153             2019-08-15           55.0                              N   
5393971             2019-06-26            0.0                              N   
5393773             2019-12-27           35.0                              N   

                 Assembly Date Attorney/Representative  Average Weekly Wage  \
Claim Identifier                                                              
5393991             2020-01-02                       Y                  0.0   
5394018             2020-01-02                       Y                  0.0   
5394153             2020-01-02                       N                  0.0   
5393971             2020-01-02                       N                  0.0   
5393773             2020-01-02                       N                  0.0   

                  Birth Year C-2 Date    C-3 Date  \
Claim Identifier                                    
5393991               1987.0      NaN  2019-12-26   
5394018               1975.0      NaN  2019-12-27   
5394153               1964.0      NaN  2019-12-31   
5393971                  0.0      NaN         NaN   
5393773                  0.0      NaN  2019-12-30   

                                  Carrier Name  ... WCIO Cause of Injury Code  \
Claim Identifier                                ...                             
5393991                   STATE INSURANCE FUND  ...                       NaN   
5394018                CHARTER OAK FIRE INS CO  ...                       NaN   
5394153                   STATE INSURANCE FUND  ...                       NaN   
5393971           AMERICAN ZURICH INSURANCE CO  ...                       NaN   
5393773            SELECTIVE INSURANCE COMPANY  ...                       NaN   

                 WCIO Cause of Injury Description WCIO Nature of Injury Code  \
Claim Identifier                                                               
5393991                                       NaN                        NaN   
5394018                                       NaN                        NaN   
5394153                                       NaN                        NaN   
5393971                                       NaN                        NaN   
5393773                                       NaN                        NaN   

                 WCIO Nature of Injury Description WCIO Part Of Body Code  \
Claim Identifier                                                            
5393991                                        NaN                    NaN   
5394018                                        NaN                    NaN   
5394153                                        NaN                    NaN   
5393971                                        NaN                    NaN   
5393773                                        NaN                    NaN   

                 WCIO Part Of Body Description Zip Code  Agreement Reached  \
Claim Identifier                                                             
5393991                                    NaN    11411                0.0   
5394018                                    NaN    07621                0.0   
5394153                                    NaN    12306                0.0   
5393971                                    NaN    11801                0.0   
5393773                                    NaN    12150                0.0   

                      WCB Decision Number of Dependents  
Claim Identifier                                         
5393991           Not Work Related                  4.0  
5394018           Not Work Related                  5.0  
5394153           Not Work Related                  5.0  
5393971           Not Wo

#### Check for unique values in each independent variable 

In [150]:
for column in train_data_cleaned.columns:
    unique_values = train_data_cleaned[column].unique()
    unique_values_num = train_data_cleaned[column].nunique()
    print(f"Unique values in '{column}':")
    print(unique_values)
    print(f"Number of unique values in '{column}':")
    print(unique_values_num)
    
    print()

Unique values in 'Accident Date':
['2019-12-30' '2019-08-30' '2019-12-06' ... '2000-05-24' '2007-05-01'
 '1980-05-17']
Number of unique values in 'Accident Date':
5539

Unique values in 'Age at Injury':
[ 31.  46.  40.  61.  67.  48.  33.  55.  20.  21.  51.  62.  35.  54.
  32.  34.  38.  30.  36.  39.  44.  56.  29.  60.  49.  50.  18.  58.
  24.  45.  53.  52.  47.  25.  23.  22.  64.  70.  41.  69.  57.  27.
  42.  77.  72.  19.  28.  63.  65.  59.  43.  26.  74.  66.  76.  68.
  37.   0.  16.  71.  82.  75.  17.  78.  73.  89.  85.  84.  81.  87.
  80.  11.  88.  79.  15.  83.  86.  90.   1.  10. 102.  94. 117.  14.
 113.  95.  92. 104.   5.   8. 115. 110.  91.  96.  99.  12.  13.  93.
 101.   7. 111.   9.  97. 112. 109. 100.  98. 114.]
Number of unique values in 'Age at Injury':
108

Unique values in 'Alternative Dispute Resolution':
['N' 'Y' 'U']
Number of unique values in 'Alternative Dispute Resolution':
3

Unique values in 'Assembly Date':
['2020-01-01' '2020-01-02' '2020-01-

#### All listed Variables - summary 

1. **Accident Date**: 0.64% NaN and dates (5539 unique values)

2. **Age at Injury**: 0% NaN and no strange values (108 unique values) (there are some values >100)
3. **Alternative Dispute Resolution**: 0% NaN and no strange values (3 unique values)
4. **Assembly Date**: 0% NaN and dates (897 unique values)
5. **Attorney/Representative**: 0% NaN and no strange values (2 unique values)
6. **Average Weekly Wage**: around 5% NaN and a lot of unique values (120024 unique values)
7. **Birth Year**: 5% Nan and one 25081 strange values "0" (Birth Year cannot be 0) (107 unique values - 0 included)
8. **C-2 Date**: 2.5% NaN and dataes (2475 unique values)
9. **C-3 Date**: 67% NaN and dates (1648 unique values)
10. **Carrier Name**: 0% NaN and name of carrier (2046 unique values) 
11. **Carrier Type**: 0% NaN and no strange values (8 unique values)
12. **Claim Injury Type** - target variable
13. **County of Injury**: 0% NaN and there are 1196 "UNKNOWN" values (63 unique values)
14. **COVID-19 Indicator**: 0% NaN and no strange values (2 unique values)
15. **District Name**: 0% NaN and no strange values (8 unique values)
16. **First Hearing Date**: 73.7% NaN and dates (1094 unique values)
17. **Gender**: 0% NaN and male, female, non-binary and unspecified (4 unique values)
18. **IME-4 Count**: 76.9% NaN and no strange values (41 unique values)
19. **Industry Code**: 1.73% NaN and no strange values (24 unique values)
20. **Industry Code Description**: 1.73% NaN and descripion (20 unique values)
21. **Medical Fee Region**: 0% NaN and one strange value 'UK' (33472 such values) other are numbers (4 unique values)
22. **WCIO Cause of Injury Code**: 2.724% NaN and no strange values (77 unique values) 
23. **WCIO Cause of Injury Description**: 2.724% NaN and no strange values (74 unique values)
24. **WCIO Nature of Injury Code**: 2.727% NaN and no strange values (56 unique values)
25. **WCIO Nature of Injury Description**: 2.727% NaN and no strange values (56 unique values)
26. **WCIO Part Of Body Code**: 2.98% NaN and 42011 "-9" values (probably the should be +9) (57 unique values)
27. **WCIO Part Of Body Description**: 2.98% NaN andone strange value: INSUFFICIENT INFO TO PROPERLY IDENTIFY - UNCLASSIFIED (54 unique values)
28. **Zip Code**: 4.98% NaN and zip code is a bit strange -> it has a lot of unique values, some zip codes starts with numbers and there are som that starts with letters; one zip code unique value is 'UNKNO". According to definition of Zip-Code: zip codes only contain a combination of numeric values.
29. **Agreement Reached**: 0% NaN and no strange values (2 unique values)
30. **WCB Decision**: 0% NaN and only 1 unique value -> variable that should be deleted it is not informative
31. **Number of Dependents**: 0% NaN and no strange values (7 unique values)



### Recommendations to discuss:
* 6 Average Weekly ge -> that variable has a lot of unique values so it is not so informative, think about creating a new variable from that with less categories
* 7 Birth Date -> that variable can be deleted 'cause we have variable age at the injury that does not have any missing value and it gives us same info as birth date
* 8 and 9 C-2/3 Date has a lot of unique values + C3 has a lot of missing values. Maybe it can be change into years only, also so many missing values in C-3 Date maybe states that Data Form C-3 has not been received.
* 10 and 11: those variables have similar information, but carier type has only 8 unique values so i would keep carrier type and drop carrier name. 
* 16 first hearing date - that variable has a lot of missing values, but those missing values indicates that the claim has not yet had a hearing held (maybe can be change to categorical variable where 1 means that first hearing date was held and 0 means that it has not been yet held) (we would be reduce in that way number of unique values)
* 18 IME-4  count: i would drop that variable since it has so many nan values 
* 19 and 20 either industry code or description should be kept: i would keep industry code
* 21 medical region UK can stands for unknown 
* 22 - 27. WCIO Columns -> we have to decide either to keep code or description: i think code is in general better (so either code or description variables have to be dropped)
* 28 zip code -> i would create divide zip codes per groups e.x one group zip hodes starts with 1, second with 2 etc. But i am not sure what to do with zip codes that strats with letters
* 30 WCB Decision -> only one unique value - definetly to drop




In [204]:
# Dropping WCB Decision variables
train_data_cleaned =train_data_cleaned.drop(['WCB Decision'], axis=1, inplace=True)

In [152]:
train_data_cleaned["Age at Injury"].value_counts().sort_index()

Age at Injury
0.0       5463
1.0          2
5.0          1
7.0          2
8.0          6
9.0          4
10.0         5
11.0         9
12.0         9
13.0         9
14.0        47
15.0       126
16.0       579
17.0      1322
18.0      3575
19.0      5960
20.0      7420
21.0      8356
22.0      9495
23.0     10649
24.0     10835
25.0     11550
26.0     12213
27.0     12679
28.0     13033
29.0     13657
30.0     14022
31.0     14041
32.0     13994
33.0     13480
34.0     13218
35.0     12885
36.0     12657
37.0     12913
38.0     12312
39.0     12190
40.0     11922
41.0     12108
42.0     11909
43.0     11511
44.0     11223
45.0     11270
46.0     11332
47.0     11539
48.0     11691
49.0     12453
50.0     12997
51.0     13486
52.0     13065
53.0     13274
54.0     13403
55.0     13305
56.0     12932
57.0     12744
58.0     12406
59.0     11571
60.0     11097
61.0     10437
62.0      8802
63.0      7487
64.0      6396
65.0      5116
66.0      3876
67.0      3070
68.0      2393
69.0      1

In [153]:
train_data_cleaned["Birth Year"].value_counts().sort_index()

Birth Year
0.0       25081
1900.0        2
1901.0       30
1902.0       28
1903.0       35
1907.0        1
1909.0        1
1910.0       39
1911.0        1
1916.0        6
1917.0        2
1920.0        5
1921.0        1
1922.0        8
1923.0        2
1924.0        4
1925.0       73
1926.0        4
1927.0        3
1928.0        4
1929.0       10
1930.0       23
1931.0       14
1932.0       22
1933.0       29
1934.0       36
1935.0       41
1936.0       70
1937.0       73
1938.0      102
1939.0      138
1940.0      148
1941.0      237
1942.0      278
1943.0      318
1944.0      386
1945.0      443
1946.0      598
1947.0      815
1948.0      916
1949.0     1062
1950.0     1381
1951.0     1683
1952.0     2050
1953.0     2689
1954.0     3479
1955.0     4284
1956.0     5420
1957.0     6551
1958.0     7803
1959.0     8746
1960.0    10034
1961.0    10640
1962.0    10998
1963.0    11632
1964.0    12279
1965.0    12168
1966.0    12249
1967.0    12101
1968.0    11953
1969.0    12353
1970.0    125

In [154]:
train_data_cleaned['County of Injury'].value_counts()

County of Injury
SUFFOLK         60430
QUEENS          58333
KINGS           53207
NASSAU          43580
BRONX           39618
ERIE            30874
NEW YORK        30014
WESTCHESTER     25754
MONROE          24071
ORANGE          17450
ONONDAGA        15511
RICHMOND        14001
ALBANY          13665
DUTCHESS        11530
ROCKLAND         9125
SARATOGA         7505
NIAGARA          7248
BROOME           6727
ONEIDA           6256
RENSSELAER       5917
ULSTER           5243
CAYUGA           4541
HERKIMER         4413
CHAUTAUQUA       4078
ONTARIO          4066
CHEMUNG          3812
OSWEGO           3741
FULTON           3581
PUTNAM           3472
ST. LAWRENCE     3355
JEFFERSON        3238
CLINTON          3053
CATTARAUGUS      2933
SULLIVAN         2862
GENESEE          2728
COLUMBIA         2582
MADISON          2435
WARREN           2268
LIVINGSTON       2111
DELAWARE         1884
WASHINGTON       1808
GREENE           1753
ALLEGANY         1743
WAYNE            1628
CHENANGO       

In [155]:
train_data_cleaned['Medical Fee Region'].value_counts()

Medical Fee Region
IV     265981
I      135885
II      85033
III     53654
UK      33472
Name: count, dtype: int64

In [156]:
train_data_cleaned["WCIO Part Of Body Code"].value_counts().sort_index()

WCIO Part Of Body Code
-9.0     42011
 10.0     7939
 11.0     7904
 12.0     1989
 13.0     3912
 14.0    12975
 15.0     2091
 16.0      803
 17.0     1602
 18.0    13765
 19.0     3905
 20.0     2848
 21.0     1332
 22.0     1103
 23.0      477
 24.0      141
 25.0     5938
 26.0       83
 30.0     2633
 31.0    11236
 32.0     9949
 33.0    15372
 34.0    18750
 35.0    31388
 36.0    36093
 37.0    12525
 38.0    37811
 39.0     2623
 40.0     1752
 41.0     8240
 42.0    51862
 43.0     1893
 44.0     7644
 45.0      402
 46.0      740
 47.0      402
 48.0     3759
 49.0      637
 50.0     1927
 51.0     6430
 52.0     4296
 53.0    47903
 54.0    11622
 55.0    24928
 56.0    18765
 57.0     2535
 58.0     1806
 60.0    14000
 61.0     6770
 62.0     1161
 63.0     1998
 64.0       53
 65.0     5311
 66.0     6103
 90.0    22608
 91.0    10950
 99.0     1249
Name: count, dtype: int64

In [157]:
train_data_cleaned["Zip Code"].value_counts().sort_index()

Zip Code
.1605       1
00000      13
00011       1
00601       1
00603       3
00623       2
00624       1
00627       1
00641       1
00656       2
00659       3
00660       1
00662       5
00669       1
00677       4
00678       1
00680       2
00685       1
00692       1
00698       1
00726       1
00727       1
00730       1
00731       1
00735       1
00737       1
00738       2
00739       1
00749       1
00751       2
00754       1
00757       1
00767       2
00772       1
00775       1
00780       1
00783       1
00791       5
00801       1
00817       1
00824       2
00857       2
00859       1
00901       1
00912       1
00925       1
00926       1
00927       1
00952       1
00956       1
00962       1
00968       1
00979       2
00982       1
01001       1
01002       4
01007       1
01013       2
01020       2
01027       1
01031       1
01034       1
01040       3
01056       3
01057       1
01060       3
01062       1
01075       1
01077       1
01083       1
01085      

#### Re-Check for duplicates 

In [158]:
# check for duplicate values on train and test dataset:
print(f"Train dataset:\n{train_data.duplicated().value_counts()}") # true: nr of duplicated rows in dataset

Train dataset:
False    575121
Name: count, dtype: int64


In [159]:
# Check the percentage of duplicates for the entire training dataset
train_data.duplicated().mean()*100
# conclusion: Not relevant for analysis, drop duplicates

0.0

In [160]:
train_data.drop_duplicates(inplace=True)
train_data.duplicated().value_counts()
# Remove duplicates

False    575121
Name: count, dtype: int64

#### Re check descriptives:

In [161]:
train_data.describe(include="all").T

count unique  \
Accident Date                         570337   5539   
Age at Injury                       574025.0    NaN   
Alternative Dispute Resolution        574025      3   
Assembly Date                         575121   1096   
Attorney/Representative               574025      2   
Average Weekly Wage                 545374.0    NaN   
Birth Year                          544947.0    NaN   
C-2 Date                              559466   2475   
C-3 Date                              187244   1648   
Carrier Name                          574025   2046   
Carrier Type                          574025      8   
Claim Injury Type                     574025      8   
County of Injury                      574025     63   
COVID-19 Indicator                    574025      2   
District Name                         574025      8   
First Hearing Date                    150797   1094   
Gender                                574025      4   
IME-4 Count                         132803.0    NaN   
Industry Code                       564068.0    NaN   
Industry Code Description             564068     20   
Medical Fee Region                    574025      5   
OIICS Nature of Injury Description       0.0    NaN   
WCIO Cause of Injury Code           558386.0    NaN   
WCIO Cause of Injury Description      558386     74   
WCIO Nature of Injury Code          558369.0    NaN   
WCIO Nature of Injury Description     558369     56   
WCIO Part Of Body Code              556944.0    NaN   
WCIO Part Of Body Description         556944     54   
Zip Code                              545388   8286   
Agreement Reached                   574025.0    NaN   
WCB Decision                          574025      1   
Number of Dependents                574025.0    NaN   

                                                                  top    freq  \
Accident Date                                              2020-03-01    1245   
Age at Injury                                                     NaN     NaN   
Alternative Dispute Resolution                                      N  571411   
Assembly Date                                              2020-03-06    1414   
Attorney/Representative                                             N  392291   
Average Weekly Wage                                               NaN     NaN   
Birth Year                                                        NaN     NaN   
C-2 Date                                                   2021-05-11    1847   
C-3 Date                                                   2021-04-21     350   
Carrier Name                                     STATE INSURANCE FUND  111144   
Carrier Type                                              1A. PRIVATE  285367   
Claim Injury Type                                         2. NON-COMP  291078   
County of Injury                                              SUFFOLK   60430   
COVID-19 Indicator                                                  N  546504   
District Name                                                     NYC  270778   
First Hearing Date                                         2022-04-15     440   
Gender                                                              M  335217   
IME-4 Count                                                       NaN     NaN   
Industry Code                                                     NaN     NaN   
Industry Code Description           HEALTH CARE AND SOCIAL ASSISTANCE  114339   
Medical Fee Region                                                 IV  265981   
OIICS Nature of Injury Description                                NaN     NaN   
WCIO Cause of Injury Code                                         NaN     NaN   
WCIO Cause of Injury Description                              LIFTING   46610   
WCIO Nature of Injury Code                                        NaN     NaN   
WCIO Nature of Injury Description                      STRAIN OR TEAR  153373   
WCIO Part Of Body Code              

#### Missing Values: (Drop/Don't drop??)

In [162]:
# Check for missing values on train dataset: 
missing_values_train = train_data.isna().sum()
missing_values_train_percentage = (train_data.isna().mean()*100).round(2)
print("Missing values\n",missing_values_train,"\n")
print("Percentage of Missing values\n",missing_values_train_percentage)

Missing values
 Accident Date                           4784
Age at Injury                           1096
Alternative Dispute Resolution          1096
Assembly Date                              0
Attorney/Representative                 1096
Average Weekly Wage                    29747
Birth Year                             30174
C-2 Date                               15655
C-3 Date                              387877
Carrier Name                            1096
Carrier Type                            1096
Claim Injury Type                       1096
County of Injury                        1096
COVID-19 Indicator                      1096
District Name                           1096
First Hearing Date                    424324
Gender                                  1096
IME-4 Count                           442318
Industry Code                          11053
Industry Code Description              11053
Medical Fee Region                      1096
OIICS Nature of Injury Description    5

#### - Unique values (search for strange values)

In [163]:
train_data.columns.values

array(['Accident Date', 'Age at Injury', 'Alternative Dispute Resolution',
       'Assembly Date', 'Attorney/Representative', 'Average Weekly Wage',
       'Birth Year', 'C-2 Date', 'C-3 Date', 'Carrier Name',
       'Carrier Type', 'Claim Injury Type', 'County of Injury',
       'COVID-19 Indicator', 'District Name', 'First Hearing Date',
       'Gender', 'IME-4 Count', 'Industry Code',
       'Industry Code Description', 'Medical Fee Region',
       'OIICS Nature of Injury Description', 'WCIO Cause of Injury Code',
       'WCIO Cause of Injury Description', 'WCIO Nature of Injury Code',
       'WCIO Nature of Injury Description', 'WCIO Part Of Body Code',
       'WCIO Part Of Body Description', 'Zip Code', 'Agreement Reached',
       'WCB Decision', 'Number of Dependents'], dtype=object)

In [164]:
train_data.shape

(575121, 32)

In [165]:
train_data["Claim Injury Type"].unique().tolist()
# Accident date
# Age at injury (0)????
# Alternative dispute resolution (U ??)
# Attorney representative (no issues)
# avg weekly wage (didn't check)
# Birth year (0), (2018 - baby)???
# carrier names - check if carrier names are weird (like general business names - walmart??)
# Carrier type: 'UNKNOWN'/ 5D. SPECIAL FUND - UNKNOWN (no issues)
# Claim injury type (cancelled ??)
#County of injury (unknown)
# Covid19 (no issues)
# district name ("Statewide")
# Gender ("U", "X")
# IME-4 Count (no issues - outlier)
# Industry code (no issues)
# Industry code description (no issues)
# Medical Fee region (UK??)
# OIICS Nature of Injury description (no values) --> removed 
# WCIO Cause of Injury Code (no issues)
# WCIO Cause of Injury Description(no issues)
# WCIO Nature of Injury Code (no issues)
# WCIO Nature of Injury Description (no issues)
# WCIO Part Of Body Code (n issues)
# WCIO Part Of Body Description ("Insufficient info to properly classify")
# Zip Code (no issues)
# Agreement Reached (no issues)
# WCB Decision (no issues)
# Number of Dependents (no issues)

['2. NON-COMP',
 '4. TEMPORARY',
 nan,
 '3. MED ONLY',
 '5. PPD SCH LOSS',
 '6. PPD NSL',
 '1. CANCELLED',
 '8. DEATH',
 '7. PTD']

#### Data types:

In [166]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 575121 entries, 5393875 to 6165075
Data columns (total 32 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   Accident Date                       570337 non-null  object 
 1   Age at Injury                       574025 non-null  float64
 2   Alternative Dispute Resolution      574025 non-null  object 
 3   Assembly Date                       575121 non-null  object 
 4   Attorney/Representative             574025 non-null  object 
 5   Average Weekly Wage                 545374 non-null  float64
 6   Birth Year                          544947 non-null  float64
 7   C-2 Date                            559466 non-null  object 
 8   C-3 Date                            187244 non-null  object 
 9   Carrier Name                        574025 non-null  object 
 10  Carrier Type                        574025 non-null  object 
 11  Claim Injury Type       

In [167]:
train_data["Age at Injury"] = train_data["Age at Injury"].astype("Int64")

In [168]:
train_data["Number of Dependents"] = train_data["Number of Dependents"].astype("Int64")

In [169]:
train_data["Birth Year"] = train_data["Birth Year"].astype("Int64")

In [170]:
train_data["WCIO Part Of Body Code"] = train_data["WCIO Part Of Body Code"].astype("Int64")

In [171]:
train_data["WCIO Nature of Injury Code"] = train_data["WCIO Nature of Injury Code"].astype("Int64")

In [172]:
train_data["WCIO Cause of Injury Code"] = train_data["WCIO Cause of Injury Code"].astype("Int64")

In [173]:
train_data["Industry Code"] = train_data["Industry Code"].astype("Int64")

In [174]:
train_data["Accident Date"] = pd.to_datetime(train_data['Accident Date'], format = '%Y-%m-%d')

In [175]:
train_data["Assembly Date"] = pd.to_datetime(train_data['C-2 Date'], format = '%Y-%m-%d')

In [176]:
train_data["First Hearing Date"] = pd.to_datetime(train_data['First Hearing Date'], format = '%Y-%m-%d')

In [177]:
train_data["C-2 Date"] = pd.to_datetime(train_data['C-2 Date'], format = '%Y-%m-%d')

In [178]:
train_data["C-3 Date"] = pd.to_datetime(train_data['C-3 Date'], format = '%Y-%m-%d')

In [179]:
train_data["IME-4 Count"] = train_data["IME-4 Count"].astype("Int64")

In [180]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 575121 entries, 5393875 to 6165075
Data columns (total 32 columns):
 #   Column                              Non-Null Count   Dtype         
---  ------                              --------------   -----         
 0   Accident Date                       570337 non-null  datetime64[ns]
 1   Age at Injury                       574025 non-null  Int64         
 2   Alternative Dispute Resolution      574025 non-null  object        
 3   Assembly Date                       559466 non-null  datetime64[ns]
 4   Attorney/Representative             574025 non-null  object        
 5   Average Weekly Wage                 545374 non-null  float64       
 6   Birth Year                          544947 non-null  Int64         
 7   C-2 Date                            559466 non-null  datetime64[ns]
 8   C-3 Date                            187244 non-null  datetime64[ns]
 9   Carrier Name                        574025 non-null  object        
 10  Carrie

In [181]:
train_data["Agreement Reached"].head()

Claim Identifier
5393875      0.0
5393091      1.0
5393889      0.0
957648180    NaN
5393887      0.0
Name: Agreement Reached, dtype: float64

In [182]:
train_data.columns.values

array(['Accident Date', 'Age at Injury', 'Alternative Dispute Resolution',
       'Assembly Date', 'Attorney/Representative', 'Average Weekly Wage',
       'Birth Year', 'C-2 Date', 'C-3 Date', 'Carrier Name',
       'Carrier Type', 'Claim Injury Type', 'County of Injury',
       'COVID-19 Indicator', 'District Name', 'First Hearing Date',
       'Gender', 'IME-4 Count', 'Industry Code',
       'Industry Code Description', 'Medical Fee Region',
       'OIICS Nature of Injury Description', 'WCIO Cause of Injury Code',
       'WCIO Cause of Injury Description', 'WCIO Nature of Injury Code',
       'WCIO Nature of Injury Description', 'WCIO Part Of Body Code',
       'WCIO Part Of Body Description', 'Zip Code', 'Agreement Reached',
       'WCB Decision', 'Number of Dependents'], dtype=object)

#### Convert birth year into age:

In [183]:
train_data["Birth Year"]= datetime.now().year - train_data["Birth Year"]

In [184]:
train_data["Claim Injury Type"].unique().tolist()

['2. NON-COMP',
 '4. TEMPORARY',
 nan,
 '3. MED ONLY',
 '5. PPD SCH LOSS',
 '6. PPD NSL',
 '1. CANCELLED',
 '8. DEATH',
 '7. PTD']

#### Dates analysis:

In [185]:
train_data["Days Between Accident_Assembly"] = train_data['Assembly Date'] - train_data['Accident Date']
print(f"maximum days between assembly: {train_data['Days Between Accident_Assembly'].max()}")
print(f"minimum days between assembly: {train_data['Days Between Accident_Assembly'].min()}") 
# strange values assembly date < accident date?? 
# #same max & min as days between accident and C2.

maximum days between assembly: 21503 days 00:00:00
minimum days between assembly: -8842 days +00:00:00


In [186]:
train_data["Days Between Accident_C2"] = train_data['C-2 Date'] - train_data['Accident Date']
print(f"maximum days between assembly: {train_data['Days Between Accident_C2'].max()}")
print(f"minimum days between assembly: {train_data['Days Between Accident_C2'].min()}") 
# strange values C2 date < accident date?? 
# #same max & min as days between accident and C2.

maximum days between assembly: 21503 days 00:00:00
minimum days between assembly: -8842 days +00:00:00


In [187]:
train_data["Days Between Accident_C3"] = train_data['C-3 Date'] - train_data['Accident Date']
print(f"maximum days between assembly: {train_data['Days Between Accident_C3'].max()}")
print(f"minimum days between assembly: {train_data['Days Between Accident_C3'].min()}")
# strange values C3 date < accident date?? 

maximum days between assembly: 18566 days 00:00:00
minimum days between assembly: -9870 days +00:00:00


In [188]:
train_data["Days Between Accident_1st_Hearing"] = train_data['First Hearing Date'] - train_data['Accident Date']
print(f"maximum days between assembly: {train_data["Days Between Accident_1st_Hearing"].max()}")
print(f"minimum days between assembly: {train_data["Days Between Accident_1st_Hearing"].min()}")
# strange values 1st_Hearing < accident date?? 

maximum days between assembly: 16373 days 00:00:00
minimum days between assembly: -576 days +00:00:00


#### Group descriptions by code: 

In [189]:
group_by_industry = train_data.groupby("Industry Code")["Industry Code Description"].unique()
group_by_industry

# Different codes with same description

Industry Code
11                      [AGRICULTURE, FORESTRY, FISHING AND HUNTING]
21                                                          [MINING]
22                                                       [UTILITIES]
23                                                    [CONSTRUCTION]
31                                                   [MANUFACTURING]
32                                                   [MANUFACTURING]
33                                                   [MANUFACTURING]
42                                                 [WHOLESALE TRADE]
44                                                    [RETAIL TRADE]
45                                                    [RETAIL TRADE]
48                                  [TRANSPORTATION AND WAREHOUSING]
49                                  [TRANSPORTATION AND WAREHOUSING]
51                                                     [INFORMATION]
52                                           [FINANCE AND INSURANCE]
53                  

In [190]:
group_by_injury = train_data.groupby("WCIO Cause of Injury Code")["WCIO Cause of Injury Description"].unique()
group_by_injury

# Different codes with same description

WCIO Cause of Injury Code
1                                       [CHEMICALS]
2                       [HOT OBJECTS OR SUBSTANCES]
3                            [TEMPERATURE EXTREMES]
4                                   [FIRE OR FLAME]
5                             [STEAM OR HOT FLUIDS]
6                    [DUST, GASES, FUMES OR VAPORS]
7                               [WELDING OPERATION]
8                                       [RADIATION]
9                               [CONTACT WITH, NOC]
10                           [MACHINE OR MACHINERY]
11                     [COLD OBJECTS OR SUBSTANCES]
12                                 [OBJECT HANDLED]
13               [CAUGHT IN, UNDER OR BETWEEN, NOC]
14                          [ABNORMAL AIR PRESSURE]
15                                   [BROKEN GLASS]
16                [HAND TOOL, UTENSIL; NOT POWERED]
17                 [OBJECT BEING LIFTED OR HANDLED]
18                   [POWERED HAND TOOL, APPLIANCE]
19                     [CUT, PUNCTURE,

In [191]:
group_by_body_part = train_data.groupby("WCIO Part Of Body Code")["WCIO Part Of Body Description"].unique()
group_by_body_part

# different codes with the same description

WCIO Part Of Body Code
-9                                                 [MULTIPLE]
10                                     [MULTIPLE HEAD INJURY]
11                                                    [SKULL]
12                                                    [BRAIN]
13                                                   [EAR(S)]
14                                                   [EYE(S)]
15                                                     [NOSE]
16                                                    [TEETH]
17                                                    [MOUTH]
18                                              [SOFT TISSUE]
19                                             [FACIAL BONES]
20                                     [MULTIPLE NECK INJURY]
21                                                [VERTEBRAE]
22                                                     [DISC]
23                                              [SPINAL CORD]
24                                             

In [192]:
# Use describe() to get the descriptive statistics for all variables for train dataset
train_data.describe(include="all").T


count unique  \
Accident Date                         570337    NaN   
Age at Injury                       574025.0   <NA>   
Alternative Dispute Resolution        574025      3   
Assembly Date                         559466    NaN   
Attorney/Representative               574025      2   
Average Weekly Wage                 545374.0    NaN   
Birth Year                          544947.0   <NA>   
C-2 Date                              559466    NaN   
C-3 Date                              187244    NaN   
Carrier Name                          574025   2046   
Carrier Type                          574025      8   
Claim Injury Type                     574025      8   
County of Injury                      574025     63   
COVID-19 Indicator                    574025      2   
District Name                         574025      8   
First Hearing Date                    150797    NaN   
Gender                                574025      4   
IME-4 Count                         132803.0   <NA>   
Industry Code                       564068.0   <NA>   
Industry Code Description             564068     20   
Medical Fee Region                    574025      5   
OIICS Nature of Injury Description       0.0    NaN   
WCIO Cause of Injury Code           558386.0   <NA>   
WCIO Cause of Injury Description      558386     74   
WCIO Nature of Injury Code          558369.0   <NA>   
WCIO Nature of Injury Description     558369     56   
WCIO Part Of Body Code              556944.0   <NA>   
WCIO Part Of Body Description         556944     54   
Zip Code                              545388   8286   
Agreement Reached                   574025.0    NaN   
WCB Decision                          574025      1   
Number of Dependents                574025.0   <NA>   
Days Between Accident_Assembly        557346    NaN   
Days Between Accident_C2              557346    NaN   
Days Between Accident_C3              183798    NaN   
Days Between Accident_1st_Hearing     149347    NaN   

                                                                  top    freq  \
Accident Date                                                     NaN     NaN   
Age at Injury                                                    <NA>    <NA>   
Alternative Dispute Resolution                                      N  571411   
Assembly Date                                                     NaN     NaN   
Attorney/Representative                                             N  392291   
Average Weekly Wage                                               NaN     NaN   
Birth Year                                                       <NA>    <NA>   
C-2 Date                                                          NaN     NaN   
C-3 Date                                                          NaN     NaN   
Carrier Name                                     STATE INSURANCE FUND  111144   
Carrier Type                                              1A. PRIVATE  285367   
Claim Injury Type                                         2. NON-COMP  291078   
County of Injury                                              SUFFOLK   60430   
COVID-19 Indicator                                                  N  546504   
District Name                                                     NYC  270778   
First Hearing Date                                                NaN     NaN   
Gender                                                              M  335217   
IME-4 Count                                                      <NA>    <NA>   
Industry Code                                                    <NA>    <NA>   
Industry Code Description           HEALTH CARE AND SOCIAL ASSISTANCE  114339   
Medical Fee Region                                                 IV  265981   
OIICS Nature of Injury Description                                NaN     NaN   
WCIO Cause of Injury Code                                        <NA>    <NA>   
WCIO Cause of Injury Description                           

#### STRANGE VALUES:

In [193]:
# Check for strange characters by NaN on train dataset:
# Strange values indicative of missing values: 
# Alternative dispute resolution (percentage)
# Birth year (0) (No issues) 
# Carrier type: 'UNKNOWN'()/ 5D. SPECIAL FUND - UNKNOWN (percentage)
# Claim injury type (cancelled ??) ------------------
# County of injury (unknown) (percentage)
# Gender ("U") (percentage)
# Medical Fee region (UK??) (percentage) 
# WCIO Part Of Body Description ("Insufficient info to properly classify") (percentage)--------------
# Zip code == 0000 (??)
# Weekly wage (0) 75% of values

#### - Check percentage of strange values in the dataset:

In [194]:
# % of UK values in medical fee region - it's significant (can't remove it)
medical_fee_count = (train_data['Medical Fee Region'] == 'UK').sum()
total_count = len(train_data)
percentage_medical_fee = (medical_fee_count / total_count) * 100
percentage_medical_fee.round(4)

5.82

In [195]:
# % of U values in alternative dispute resolution - U Gender is significant. X gender, significant??
gender_count = (train_data['Gender'] == 'U').sum()
total_count = len(train_data)  
percentage_gender_U = (gender_count / total_count) * 100
print(f"Gender U: {percentage_gender_U.round(4)}%")

gender_count = (train_data['Gender'] == 'X').sum()
total_count = len(train_data)
percentage_gender_X = (gender_count / total_count) * 100
print(f"Gender X: {percentage_gender_X.round(4)}%")

Gender U: 0.8191%
Gender X: 0.008%


In [196]:
# % of "Insufficient info to properly classify" in WCIO Part Of Body Description - It's significant!
gender_count = (train_data['WCIO Part Of Body Description'] == 'INSUFFICIENT INFO TO PROPERLY IDENTIFY - UNCLASSIFIED').sum()
total_count = len(train_data)
percentage_gender_X = (gender_count / total_count) * 100
percentage_gender_X.round(4)

0.9235

In [197]:
# % of U values in alternative dispute resolution - significant??
alternative_dispute_resolution_count = (train_data['Alternative Dispute Resolution'] == 'U').sum()  
total_count = len(train_data)  
percentage_alternative_dispute_resolution = (alternative_dispute_resolution_count / total_count) * 100
percentage_alternative_dispute_resolution

0.0008693822691224976

In [198]:
# % of unknown values in county of injury - significant?? 
county_injury_count = (train_data['County of Injury'] == 'UNKNOWN').sum()
total_count = len(train_data)  
percentage_county_injury = (county_injury_count / total_count) * 100
percentage_county_injury

0.20795623877410144

#### Define target variable

In [199]:
x = train_data.drop('Claim Injury Type', axis = 1) # axis=1 means the operation must be done on columns
y = train_data['Claim Injury Type'] 


### Define metric & non-metric features

In [200]:
metric_features = ["Age at Injury", "Average Weekly Wage", "Birth Year", "IME-4 Count", "Number_of_Dependents", "Days Between Accident_Assembly", "Days Between Accident_C2", 
                   "Days Between Accident_C3", "Days Between Accident_1st_Hearing"]

non_metric_features = ["Industry Code", "WCIO Cause of Injury Code", "WCIO Nature of Injury Code", "WCIO Part Of Body Code", "Accident Date", "Alternative Dispute Resolution",
                       "Assembly Date", "Attorney/Representative", "C-2 Date", "C-3 Date", "Carrier Name", "Carrier Type", "County of Injury", "COVID-19 Indicator",
                       "District Name", "First Hearing Date", "Gender", "Industry Code Description", "Medical Fee Region", "WCIO Cause of Injury Description", 
                       "WCIO Nature of Injury Description", "WCIO Part Of Body Description", "Zip Code", "Agreement Reached", "WCB Decision"]

